src/Data/GADT/Internal.hs

```diff
@@ -195,6 +196,72 @@ instance (GRead a, GRead b) => GRead (a :+: b) where
 -- |A class for type-contexts which contain enough information
 -- to (at least in some cases) decide the equality of types
 -- occurring within them.
+--
+-- This class is sometimes confused with 'TestEquality' from base.
+-- 'TestEquality' only checks /type equality/.
+--
+-- Consider
+--
+-- >>> data Tag a where TagInt1 :: Tag Int; TagInt2 :: Tag Int
+--
+-- The correct @'TestEquality' Tag@ instance is
+--
+-- >>> :{
+-- instance TestEquality Tag where
+--     testEquality TagInt1 TagInt1 = Just Refl
+--     testEquality TagInt1 TagInt2 = Just Refl
+--     testEquality TagInt2 TagInt1 = Just Refl
+--     testEquality TagInt2 TagInt2 = Just Refl
+-- :}
+--
+-- While we can define
+--
+-- @
+-- instance 'GEq' Tag where
+--    'geq' = 'testEquality'
+-- @
+--
+-- this will mean we probably want to have
+--
+-- @
+-- instance 'Eq' Tag where
+--    _ '==' _ = True
+-- @
+--
+-- /Note:/ In the future version of @some@ package (to be released around GHC-9.6 / 9.8) the
+-- @forall a. Eq (f a)@ constraint will be added as a constraint to 'GEq',
+-- with a law relating 'GEq' and 'Eq':
+--
+-- @
+-- 'geq' x y = Just Refl  \<=>  x == y = True
+-- 'geq' x y = Nothing    \<==  x == y = False
```

Comment:
I wrote some laws in this style, but I now realize it is fishy because `x == y` is ill-typed if we don't know the parameters match.

It's very hard to write the implication without some sort of "oracle" e.g. for `'geq' x y = Nothing` we ought to have evidence of something like "Either a =/= b, or (a = b => x =/= y)". Assuming a = b for the "right" case is approximating that oracle.

Reply:
Note that in

```
 'geq' x y = Nothing    \<==  x == y = False
 ```
 
 arrow goes only to the left. So there is no typing problem. See `checkBwdGEq` example, it's homogenous.

Reply:
Will it be clearer to rewrite them as

```haskell
geq x y = Just Refl   ⇒  x == y = True        ∀ (x :: f a) (y :: f b) 
x == y                       ⇔  isJust (geq x y)     ∀ (x, y :: f a)
```

?

Reply:
Yeah maybe, that looks nice.